### PR TITLE
Use getPrettyName; cater for legacy incorrect package names.

### PR DESCRIPTION
**Pull request by gitlost**
Closes https://github.com/wp-cli/package-command/issues/43
Partially reverts https://github.com/wp-cli/package-command/pull/49

(TLDR: Alas https://github.com/wp-cli/package-command/pull/49 had a number of issues. This PR tries to deal with them and also cater for legacy incorrect names in `composer.json`.)

Uses `Package::getPrettyName()` instead of `Package::getName()` throughout `Package_Command` (except where needed for legacy situations), as `getPrettyName()` is the actual case-sensitive name that appears in `composer.json`.

Allows and corrects for legacy incorrect names being in `composer.json`, which cause strange behaviour if present. A warning is printed on detection.

Fixes a case-sensitivity bug in `Composer\Json\JsonManipulator` re package names, which causes mixed-case package names in `require` and `repositories` to be left in `composer.json`, by adding a `caseInsensitive` arg to `addLink()`, `addSubNode()` and `removeSubNode()`. This required a full copy of `Composer\Json\JsonManipulator` (due to its use of private member variables) with adaptations.

Reverts the addition in https://github.com/wp-cli/package-command/pull/49 of an outwardly visible `pretty_name`, as having two names only confuses things and there really can't be BC issues as always outputting lowercase name is wrong. For BC with `package-command` 1.0.8, still sets `pretty_name` so it won't fail if specified.

Moves the mismatch code to its own function so it can be called to deal with the shortened form of git url also. Adds some error handling and expands the mismatch message with explicit before and after names.

Also copies the phpunit test for `JsonManipulator` from Composer, with some additions.

### CODE REVIEW

**Comment by gitlost** [comment]
Re build fail there seems to be issues with github.com at the mo (although the [status](https://status.github.com/messages) isn't showing anything yet), or it could be that the `GITHUB_TOKEN` stuff https://github.com/wp-cli/package-command/pull/47 needs further work...

**Comment by danielbachhuber** [comment]
Restarted build

**Comment by gitlost** [comment]
Failed again so github.com issues can be ruled out I think and it's some issue with the `GITHUB_TOKEN`/`COMPOSER_AUTH` stuff added in https://github.com/wp-cli/package-command/pull/47 not working fully.

**Comment by danielbachhuber** [comment]
@gitlost Yes, but not in the way you think:

image

**Comment by gitlost** [comment]
Ha excellent!

**Comment by gitlost** [comment]
Ah it would fail randomly locally too - will just remove it...

**Comment by danielbachhuber** [comment]
> What would you think of adding a `@local-only` tag or similar to `ci/behat-tags.php` eg

I'm not opposed and don't have a better suggestion, if you think this is the best path forward.

However, I'm not sure of the utility of an integration test that never runs in the CI environment.

**Comment by danielbachhuber** [comment]
Requested a review from @schlessera because I'd like his eyes on it specifically.

File: src/Package_Command.php [hunk 1]
```diff
@@ -952,6 +950,37 @@ private function is_git_repository( $package ) {
 	}
 
 	/**
+	 * Check that `$package_name` matches the name in the repo composer.json, and return corrected value if not.
+	 */
+	private function check_git_package_name( $package_name ) {
+		// Generate raw git URL of composer.json file.
+		$raw_content_url = 'https://raw.githubusercontent.com/' . $package_name . '/master/composer.json';
```
Comment (schlessera) [inline src/Package_Command.php]: I think we still have a problem here.
The branch `master` is hardcoded, so the code will pull in the wrong `composer.json` file if a different branch was requested.
Not a bug with this PR, but rather a bug we already had in the previous iteration.

Comment (gitlost) [reply src/Package_Command.php]: Ah right so `check_git_package_name()` needs to take a `$version` arg as well - should this be a new issue/PR or deal with it here?

Comment (schlessera) [reply src/Package_Command.php]: I'd say a new PR, as the bug is already in there. This way, we can concentrate on the issue at hand first, and deal with the other bug separately.

Comment (schlessera) [reply src/Package_Command.php]: I'll merge this one and create a new issue for that bug.

Comment (schlessera) [reply src/Package_Command.php]: See https://github.com/wp-cli/package-command/issues/51

**Pull request merged**